okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -636,6 +639,9 @@ private Socket createHttpProxySocket(InetSocketAddress address, InetSocketAddres
             socketFactory.createSocket(proxyAddress.getHostName(), proxyAddress.getPort());
       }
       sock.setTcpNoDelay(true);
+      // A socket timeout is needed because lost network connectivity while reading from the proxy,
+      // can cause reading from the socket to hang.
+      sock.setSoTimeout(proxySocketTimeout);
```

Comment:
If a read() fails, the socket won't be closed. Seems we need to move `Socket sock` outside of the try and add a `GrpcUtil.closeQuietly(sock)` to the catch.